src/v/cloud_topics/level_one/metastore/tests/garbage_collector_test.cc

```diff
@@ -0,0 +1,262 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_topics/level_one/common/fake_io.h"
+#include "cloud_topics/level_one/metastore/garbage_collector.h"
+#include "cloud_topics/level_one/metastore/simple_stm.h"
+#include "cloud_topics/level_one/metastore/state_update.h"
+#include "config/property.h"
+#include "gtest/gtest.h"
+#include "model/fundamental.h"
+#include "raft/tests/raft_fixture.h"
+#include "storage/record_batch_builder.h"
+
+using namespace cloud_topics::l1;
+
+namespace {
+using o = kafka::offset;
+using ts = model::timestamp;
+ss::abort_source never_abort;
+ss::logger test_log("gc-test");
+} // namespace
+
+class GarbageCollectorTest : public raft::stm_raft_fixture<simple_stm> {
+public:
+    stm_shptrs_t create_stms(
+      raft::state_machine_manager_builder& builder,
+      raft::raft_node_instance& node) override {
+        return builder.create_stm<simple_stm>(
+          test_log, node.raft().get(), config::mock_binding(5s));
+    }
+
+    model::topic_id_partition make_tp(int i) {
+        return model::topic_id_partition::from(
+          fmt::format("deadbeef-aaaa-0000-0000-000000000000/{}", i));
+    }
+
+    ss::future<>
+    add_objects(simple_stm* stm, int partitions_count, int last_offset) {
+        chunked_vector<new_object> new_objects;
+        term_state_update_t terms;
+
+        for (int i = 0; i < partitions_count; ++i) {
+            auto tp = make_tp(i);
+            auto oid = create_object_id();
+
+            new_object obj;
+            obj.oid = oid;
+            obj.footer_pos = 1000;
+            obj.object_size = 1500;
+            obj.extent_metas[tp.topic_id][tp.partition] = new_object::metadata{
+              .base_offset = o{0},
+              .last_offset = o{last_offset},
+              .max_timestamp = ts{10000},
+              .filepos = 0,
+              .len = 500,
+            };
+            new_objects.push_back(std::move(obj));
+            terms[tp].emplace_back(
+              term_start{.term_id = model::term_id{0}, .start_offset = o{0}});
+
+            auto file_res = co_await _io.create_tmp_file();
+            ASSERT_TRUE_CORO(file_res.has_value());
+            auto ostream = co_await (*file_res)->output_stream();
+            co_await ostream.close();
+            auto put_res = co_await _io.put_object(
+              oid, file_res->get(), &never_abort);
+            ASSERT_TRUE_CORO(put_res.has_value());
+            co_await file_res->get()->remove();
+        }
+        auto sync_res = co_await stm->sync(10s);
+        ASSERT_TRUE_CORO(sync_res.has_value());
+
+        auto update_res = add_objects_update::build(
+          stm->state(), std::move(new_objects), std::move(terms));
+        ASSERT_TRUE_CORO(update_res.has_value());
+        storage::record_batch_builder builder(
+          model::record_batch_type::l1_stm, model::offset{0});
```

Comment:
FYI: we have a new `model::batch_builder` now that's a little simpler to use IMO